sdk/storage/azure-storage-common/src/main/java/com/azure/storage/common/implementation/UploadBufferPool.java

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.storage.common.implementation;
+
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.LinkedBlockingQueue;
+
+/**
+ * This type is to support the implementation of buffered upload only. It is mandatory that the caller has broken the
+ * source into ByteBuffers that are no greater than the size of a chunk and therefore a buffer in the pool. This is
+ * necessary because it upper bounds the number of buffers we need for a given call to write() to 2. If the size of
+ * ByteBuffer passed into write() were unbounded, the pool could stall as it would run out of buffers before it is able
+ * to return a result, and if it is unable to return, no data can be uploaded and therefore no pools returned.
+ *
+ * It is incumbent upon the caller to return the buffers after an upload is completed. It is also the caller's
+ * responsibility to signal to the pool when the stream is empty and call flush to return any data still sitting in the
+ * pool.
+ *
+ * Broadly, the workflow of this operation is to chunk the source into reasonable sized pieces. On each piece, one
+ * thread will call write on the pool. The pool will grab a buffer from the queue to write to, possibly waiting for one
+ * to be available, and either store the incomplete buffer to be filled on the next write or return the filled buffer to
+ * be sent. Filled buffers can be uploaded in parallel and should return buffers to the pool after the upload completes.
+ * Once the source terminates, it should call flush.
+ *
+ * RESERVED FOR INTERNAL USE ONLY
+ */
+public final class UploadBufferPool {
+    private final ClientLogger logger = new ClientLogger(UploadBufferPool.class);
+
+    /*
+    Note that a blocking on a synchronized object is not the same as blocking on a reactive operation; blocking on this
+    queue will not compromise the async nature of this workflow. Fluxes themselves are internally synchronized to ensure
+    only one call to onNext happens at a time.
+     */
+    private final BlockingQueue<ByteBuffer> buffers;
```

Comment:
Because this class is blocking, and because it will block for significant amounts of time given we're handling buffers with it, we should wrap it in a reactive data structure using the tips [described here](https://projectreactor.io/docs/core/release/reference/#faq.wrap-blocking) for wrapping synchronous, blocking calls.

Reply:
Just cause this class was a straight up copy paste from what exists in blobs, and it shouldnt affect any public APIs, I'll create an issue for this

Reply:
https://github.com/Azure/azure-sdk-for-java/issues/8382